diff --git a/config/meson.build b/config/meson.build
index 7134e80e8d..d31bf0db86 100644
--- a/config/meson.build
+++ b/config/meson.build
@@ -197,7 +197,7 @@ if fdt_dep.found() and cc.has_header('fdt.h')
     dpdk_extra_ldflags += '-lfdt'
 endif
 
-libexecinfo = cc.find_library('libexecinfo', required: false)
+libexecinfo = cc.find_library('libexecinfo', required: false, dirs: ['/home/mfw/openwrt/staging_dir/target-x86_64_musl/root-x86/usr/lib64'])
 if libexecinfo.found() and cc.has_header('execinfo.h')
     add_project_link_arguments('-lexecinfo', language: 'c')
     dpdk_extra_ldflags += '-lexecinfo'
