--- a/setup.py	2018-11-02 17:21:24.023564902 +0100
+++ b/setup.py	2018-11-02 17:21:58.004425606 +0100
@@ -29,7 +28,6 @@
     name="jsonschema",
     packages=["jsonschema", "jsonschema.tests"],
     package_data={"jsonschema": ["schemas/*.json"]},
-    setup_requires=["vcversioner>=2.16.0.0"],
     extras_require=extras_require,
     author="Julian Berman",
     author_email="Julian@GrayVines.com",
@@ -39,5 +37,5 @@
     long_description=long_description,
     url="http://github.com/Julian/jsonschema",
     entry_points={"console_scripts": ["jsonschema = jsonschema.cli:main"]},
-    vcversioner={"version_module_paths" : ["jsonschema/_version.py"]},
+    version=os.environ.get('PKG_VERSION')
 )
